[59845] Update Datepicker for automatic scheduling mode

HDinger opened the pull request:
# Ticket
https://community.openproject.org/projects/stream-planning-and-reporting/work_packages/59845/activity

# What are you trying to accomplish?
Primerise the datepicker dialog

**ToDo**
- [x] Render a Primer Banner with different texts depending on the data
- [x] There is a new segmented control "Scheduling mode"
- [x] The inputs are Primer Text fields
  - [x] There is a "today" link below
- [x] The dialog content (text fields) update automatically:
  - [x] when the calendar is changed
  - [x] when the text field values are changed manually
- [x] Handle special cases
  - [x] Milestones
  - [x] WorkPackage create form
  - [x] mobile
- [x] Take care the the UX behaves as before
  - [x] Focus indicates the selected field
  - [x] special logic when start date is larger than end date
  - [x] Increasing the duration will always increase the end date


## Screenshots
![Dec-18-2024 09-26-28](https://github.com/user-attachments/assets/f60a2003-9472-4077-8866-1b012b7641ac)



# What approach did you choose and why?
Moving the datepicker dialog closer towards Primer. Two things will remain however in angular:
* the modal itself, and
* the flatpickr Instance. 

The modal is so closely coupled to the edit field that it will be very hard to get rid of that. Flatpickr is an external library which will not get rid of for now.

# Merge checklist

- [ ] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

HDinger committed:
Start on using Primer components inside the datepicker modal

HDinger committed:
Remove unused angular components which are going to be replaced by ViewComponents

HDinger committed:
Use existing fields for now to get the modal working again

HDinger committed:
Render the form from rails and only the actual datepicker in angular.

HDinger committed:
Extract logic to update the edit field after the turboFrame was submitted into a directive so that it can be used for the datepicker as well

HDinger committed:
Extract logic of the progress modal to be reused for the datepicker modal as well

HDinger committed:
Update the dialog when the "ignore_non_working_days" checkbox is checked

HDinger committed:
Update the dialog content when chaning the scheduling mode

HDinger committed:
Cancel the editField correctly when canceling the dialog

HDinger committed:
First try on updating the datepicker when the fields above change

HDinger committed:
Correctly save the sheduling mode and duration

HDinger committed:
Preserve values when updating the content via turboStream

HDinger committed:
Disable the form in automatic scheduling mode

HDinger committed:
Show additional tabs in datepicker dialog

HDinger committed:
Do not update the fields so often

HDinger committed:
Correctly parese the duration

HDinger committed:
Revert commit d2a6ba2a04fe038f4f008214997e931091971cea

HDinger committed:
Improve event handling when datepicker field change so that also non-user changes are tracked

HDinger committed:
Forbid editing the relations in the datepicker

HDinger committed:
Take care that something is shown when opening the duration field

HDinger committed:
Hack some roundtrip between the datpicker and the fields above (wip)

HDinger committed:
Allow duration to be editable in automatic mode

HDinger committed:
Avoid code repetition

HDinger committed:
Extract doubled code into a method

HDinger committed:
Please rubocop and eslint

HDinger committed:
Handle Milestones in new datepicker modal

HDinger committed:
If the WP is automatically scheduled but there is no predecessor, we show a blankslate

HDinger committed:
Fine tune text and spaces

HDinger committed:
Datepicker for WorkPackage Create case

HDinger committed:
Disbale saving when scheduling is not possible

HDinger committed:
Provide Blankslates for empty tabs

HDinger committed:
Avoid that the Dialog is jumping around when switching tabs and the content size changes

HDinger committed:
Do not show the banner when there are no relations in manual mode

HDinger committed:
Start on mobile layout for Datepicker

HDinger committed:
Extract code into a separate component for better structure

HDinger committed:
Hide the UnderlineNav on mobile

HDinger committed:
Take care that the datepicker scrolls on mobile

HDinger committed:
Add "Today" link for Datepicker fields

HDinger committed:
Remove unused sass styles

HDinger committed:
Start to re-add some test selectors and adapt tests to new structure

HDinger committed:
Always allow editing the duration

HDinger committed:
Set the test_selector correctly

HDinger committed:
Define the currently active field correctly and respect that when selecting a date from the datepicker

HDinger committed:
Get tests running by waiting for the controller to be loaded before starting to change things. Further some selectors have been fixed

HDinger committed:
Disable duration and NonWorkingDays as specified

HDinger committed:
Adapt some more test selectors

cbliard committed:
Do not display 'Today' link for duration

cbliard committed:
Fix rubocop complains

cbliard committed:
Ensure reinitialization of datepicker on change events is done only once

HDinger committed:
Fix some more tests

cbliard committed:
Fix progress popover features tests

Debounce value was changed in the app part, but was not updated in the
test.

Also, changed the way the field is focused to avoid the select to be
opened (because it could stay opened sometimes). And for input fields
being focused, it required to also move the carte to the end of the
input before clearing it. It was not needed before because we used to
click in the field to give focus, and clicking also moves the caret to
the end (most of the time).

cbliard committed:
Fix date conversion when date is not set

HDinger committed:
Take care that creation of Milestones is working correctly

cbliard committed:
Disable a test until we have a 'clear' button on duration field

HDinger committed:
Show date form when the WorkPackage is in automatic scheduling mode and has children

cbliard committed:
Improve datepicker behavior when choosing dates

Make it so that selecting dates in the flatpicker calendar no longer
displays "this date should be before/after the other".

HDinger committed:
Add trailing unit for duration field

HDinger committed:
Add special logic for edge case in which only one date is given and the user wants to modify exactly that

HDinger committed:
Switch the focused date field if one of the values is empty

HDinger committed:
Only change the focus when the triggered field was the combinedField. Otherwise we respect the users choice

HDinger committed:
Move test to cuprite in order to wait for the idle network

HDinger committed:
Re-add selenium test

HDinger committed:
Fix some more test selectors

HDinger committed:
Fix some more test selectors

HDinger committed:
Correct test expectations for Banner messages in the datepicker

cbliard committed:
Preload DatePicker PreviewController to make tests pass

When in tests the date picker is opened to click on a day of the
flatpickr, the controller is not loaded yet and some events are missed,
making the test fail.

The controller is not ready yet because it needs to be loaded
dynamically by the OpApplicationController. By registering it in the
setup.ts, it is loaded before the date picker modal is opened, and the
controller is connected before a day is clicked, making (hopefully) the
test pass.

cbliard committed:
Fix input selector for milestone in DateEditField

HDinger committed:
Remove unused services

HDinger committed:
Fix banner expectation text and the condition for it to be shown

cbliard committed:
Improve month selection for date picker tests

After selecting a month, `#select_month` returns the index of the
selected month. It previously worked well only for single-month
datepickers. For a multiple months date picker, it was failing because
the selector is invalid (so it waited 5 seconds for nothing), then it
was retried, the current month was calculated, and that was the value
returned as the 2 conditions were not met. So in the end it was
returning the correct value, but it was also doing one useless retry.

This commit introduces `#current_month_index` to return the index of the
current month. This one works for both single and multi date pickers. It
is used by `#select_month` to calculate the number of months to move
forward or backward.

This should lead to some faster tests.

cbliard committed:
Fix test

The date picker updates its calendar more frequently than needed. When
non-working days is toggled, the flatpickr is reinitialized. If the date
is selected before this reinitialization is finished, it selects the
date, then the reinitization happens so it displays the current date,
and then we do some assertions on dates that are no longer displayed.
That's why the test was failing.

cbliard committed:
Better test assertions and documentation

Do not set expectations on the day being waited for. Could be disabled
or not, that's not this test purpose, that only to wait for a refresh.

And use default wait time. 20 seconds is way to much to wait.

HDinger committed:
Correctly show all WPs in the linked relation Gantt view and adjust some tests

HDinger committed:
Only link visible relations in the banner

HDinger committed:
Add test for banner if the predecessors have a large gap or are overlaping

HDinger committed:
Improve behaviour when the duration field is selected

HDinger committed:
Add missing values for test scenarios

HDinger committed:
Adapt test to new behavior

cbliard committed:
Display error in date picker for invalid duration

If duration is negative or garbage, an error is displayed in the date
picker. It will not display any derivation-related errors for start and
due dates. For instance if start date is set, due date is not set, and
duration is invalid, then it won't display "can not be set to null as
start date and duration are known" error for due date.

lindenthal commented:
![image](https://github.com/user-attachments/assets/fb6feb94-cd61-4393-899c-67988a42358c)

hi @HDinger,

Is it possible to have the date fields taking less space on mobile? 

![image](https://github.com/user-attachments/assets/1a7909bb-6db4-4969-b99b-d1a95bf37a2e)

HDinger committed:
Ensure that only two fields at most are send to the backend, otherwise we can easily create errorneous data. Therefore the logic of the progress controller was moved to the parent controller and abstracted as it is doing the exact same thing.

HDinger commented:
> ![image](https://private-user-images.githubusercontent.com/1019287/407039157-fb6feb94-cd61-4393-899c-67988a42358c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgxNDA1MDUsIm5iZiI6MTczODE0MDIwNSwicGF0aCI6Ii8xMDE5Mjg3LzQwNzAzOTE1Ny1mYjZmZWI5NC1jZDYxLTQzOTMtODk5Yy02Nzk4OGE0MjM1OGMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMjlUMDg0MzI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZGJmZDAzYzZjOTQ0MjgyMDU1MDVmZDJhOWE5ZGQ1YjkxODVmNWE0ZDVmMmQ1ZDc4M2E4ZjNjOTU1NWIyN2M3MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.MtOiON_seFZDJsxyzxcbEHWRu2eXoFZhNhSrNcV8wTE)
> 
> hi @HDinger,
> 
> Is it possible to have the date fields taking less space on mobile?
> 
> ![image](https://private-user-images.githubusercontent.com/1019287/407039757-1a7909bb-6db4-4969-b99b-d1a95bf37a2e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgxNDA1MDUsIm5iZiI6MTczODE0MDIwNSwicGF0aCI6Ii8xMDE5Mjg3LzQwNzAzOTc1Ny0xYTc5MDliYi02ZGI0LTQ5NjktYjk5Yi1kMWE5NWJmMzdhMmUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMjlUMDg0MzI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmFlMDhmZDBhYjBjZDBkNGZlMzNhYWJjYzIwMzk2NGYzYjY2YWM4OTA2NDE3MmI1ODFlYzc0YzA1YzAwMDA0OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.f3jS4CU6THLLUwVEHnED8Z4kgAPoouDIuo0Kcb21OyU)

Hi @lindenthal 

there are still some ongoing discussions about the mobile layout in [this ticket](https://community.openproject.org/work_packages/61051). I added your suggestion there as well 👍

HDinger committed:
Add icon to remove start or due date

cbliard committed:
Fix clicking on start date after due date in date picker

When doing so, it was comparing the start date given as tag attribute to
the clicked date. It was comparing a string '2025-01-30' to a date
object.

Now there are fields for start date and due date attributes, and fields
for the actual value as `Date`. The duration field is also cleared in
this situation.

Updated a test to click on a calendar date instead of using the field
directly so that it is covered by tests.

cbliard committed:
[60670] Update datepicker tests according to the new behavior

Some scenarios have been described in https://community.openproject.org/wp/60670

Some of them are still not implemented. 5 of them are still failing:
  - scenarios 13x
  - scenario 22a
  - scenario 23a

scenario 6a relies on the clear button being available, which is not the
case yet, so it's disabled.

cbliard committed:
Remove no more relevant test

The duration does not have any "clear" button anymore.

cbliard committed:
refactor: Make test terminology match UI terminology

cbliard committed:
Restore progress preview controller as it was before

The date picker preview controller and the progress preview controller
were abstracted into a super class. It's too hard to make deep
modifications to the date picker preview controller while keeping the
compatibility with the progress preview controller.

For instance, we break the progress values being formatted on blur
without noticing. This commit adds a test just for this.

This commit undoes some changes made in 5429b8b (commit title is
"Extract logic of the progress modal to be reused for the datepicker
modal").

Once the date picker progress controller works as expected, we'll see
which part can be reused and extracted into an abstraction.

cbliard committed:
Restore debounce of 200ms in progress preview controller (was 100ms)

cbliard committed:
Handle dates swapping when editing dates fields directly

This logic was in the date picker component, meaning it was triggered
when picking another date, but not when editing dates in the input
fields directly.

This commit moves the logic to the preview controller, and removes
almost all responsibilities from the date picker component:

- It sends `date-picker:flatpickr-dates-changed` events to the
  controller when new dates are picked
- It receives `date-picker:flatpickr-set-values events from the
  controller to update its dates, mode, and display of non-working days.
  When the update is done, it does it without reinitializing itself.

Some logic still remains in the date picker component for its
initialization which is based on its HTML attributes.

The controller also changed how it gets the dates values after a turbo
morphing: the values are read from HTML inputs once the morph rendering
is finished.

There are also 2 different debounce for the preview: one of 200ms when
the user is typing in the input fields, and one of 0ms when the user is
clicking on a date in the calendar. In that case the debounce is not
needed as the date is set immediately.

cbliard committed:
Update test to new behavior

If the dates have not changed, the date picker is not reinitialized and
its dates are not set again. So switching "Working-days only" checkbox
no longer shows the current date again and it's not needed for the test
to wait for the current date to be displayed and switch back to the
correct date.

cbliard committed:
Fix missing children and ancestors visibility control

And fix rubocop Metrics/AbcSize offense.

cbliard committed:
Fix erb_lint offenses

cbliard committed:
Display ghost relations properly in date picker tabs
